### ISSUE TITLE
HttpAssertion.WithRequestBody throws ArgumentNullException when any logged calls don't have a RequestBody

### ISSUE THREAD

**Issue by andycmaj** [tmenier/Flurl#169]
error:
```
Value cannot be null.
Parameter name: input
   at System.Text.RegularExpressions.Regex.IsMatch(String input)
   at System.Text.RegularExpressions.Regex.IsMatch(String input, String pattern)
   at System.Linq.Enumerable.WhereListIterator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Flurl.Http.Testing.HttpCallAssertion.With(Func`2 match) in C:\projects\flurl\src\Flurl.Http.Shared\Testing\HttpCallAssertion.cs:line 224
   at Flurl.Http.Testing.HttpCallAssertion.WithRequestBody(String bodyPattern) in C:\projects\flurl\src\Flurl.Http.Shared\Testing\HttpCallAssertion.cs:line 167
   at TheChunnel.Activities.Marketing.Tests.Webhooks.SendToEndpointCommand.SendToEndpointCommandTestContext`2.AssertMockRestClientCall(Int32 timesCalled, IRestRequest restRequest) in \\Mac\appature\git\TheChunnel\test\TheChunnel.Activities.Marketing.Tests\Webhooks\SendToEndpointCommand\SendToEndpointCommandTestContext.cs:line 266
   at TheChunnel.Activities.Marketing.Tests.Webhooks.SendToEndpointCommand.SendToEndpointCommandTestContext`2.AssertWebhookTransmissionResult(UInt32 webhookId, IRestRequest expectedRestRequest, String expectedEndpoint, WebhookTransmissionStatus expectedOutputStatus, IRestResponse restResponse, String expectedErrorMessageSubstring) in \\Mac\appature\git\TheChunnel\test\TheChunnel.Activities.Marketing.Tests\Webhooks\SendToEndpointCommand\SendToEndpointCommandTestContext.cs:line 249
   at TheChunnel.Activities.Marketing.Tests.Webhooks.SendToEndpointCommand.V1.Test.ValidInput_ShouldSendToEndpoint_AndShouldRecordResult() in \\Mac\appature\git\TheChunnel\test\TheChunnel.Activities.Marketing.Tests\Webhooks\SendToEndpointCommand\V1\Test.cs:line 74
```

this seems to happen when i call 
```
                Http.ShouldHaveMadeACall()
                    .WithRequestBody(restRequest.Body)
```

when any of `HttpTest.CallLog` do not have a `RequestBody`.

<img width="1920" alt="screencapture at mon feb 20 14 56 21 pst 2017" src="https://cloud.githubusercontent.com/assets/97470/23144553/445e6914-f77d-11e6-8c5d-82ed29447959.png">
<img width="1920" alt="screencapture at mon feb 20 14 55 57 pst 2017" src="https://cloud.githubusercontent.com/assets/97470/23144552/445d84cc-f77d-11e6-9c26-b9c5bfe291a5.png">

seems to be because this Assertion looks at `RequestBody`, and none seem to be recorded even when using `SendAsync`.

https://github.com/tmenier/Flurl/blob/master/src/Flurl.Http.Shared/Testing/HttpCallAssertion.cs#L167

here's the call i'm making that doesn't seem to have `RequestBody` set:

```
requestTask = client
                        .SendAsync(
                            requestMethod,
                            new StringContent(webhookHit.ExtraParameters, null, webhookHit.ContentType)
                        );
```

**Comment by andycmaj** [comment]
Here's a simple test that repros this:

<img width="980" alt="screencapture at mon feb 20 15 03 59 pst 2017" src="https://cloud.githubusercontent.com/assets/97470/23144643/e0de6a00-f77d-11e6-90a3-d3dfc2dda335.png">


```
using (var httpTest = new HttpTest())
{
	"http://foo.bar".SendAsync(HttpMethod.Post, new StringContent("foo", null, "text/plain")).Wait();
	
	httpTest.CallLog[0].RequestBody.Dump("Request Body");
	
	httpTest.ShouldHaveMadeACall().WithRequestBody("foo");
}
```

**Comment by andycmaj** [comment]
this seems like 2 bugs actually:
1. `RequestBody` does not seem to get set on logged `HttpCalls` during `HttpTest`s.
2. `WithRequestBody` should not expect EVERY call in the call log to have a populated request body. If you are mixing GETs and POSTs, this will cause exceptions.

**Comment by tmenier** [comment]
Thanks for reporting. I'll look into fixing soon.

**Comment by tmenier** [comment]
Here's some info on why this is happening.

TL;DR

Instead of doing this:

`"http://foo.bar".SendAsync(HttpMethod.Post, new StringContent("foo", null, "text/plain"))`

do this:

"http://foo.bar".PostStringAsync("foo")

In the first instance, `RequestBody` is null because [StringContent](https://msdn.microsoft.com/en-us/library/system.net.http.stringcontent.aspx) (ironically) doesn't expose the string you set in the constructor via any property, it only exposes it as a steam. This is exactly where Flurl's [CapturedStringContent](https://github.com/tmenier/Flurl/blob/master/src/Flurl.Http.Shared/Content/CapturedStringContent.cs) comes into play. It extends `StringContent` by simply exposing that string as a property called `Content`, which the call log then uses for `RequestBody`. Flurl's `PostStringAsync` uses `CapturedStringContent`, which is why the alternative above will work. You can also create a `CapturedStringContent` explicitly for use with `SendAsync` if you want.

That said, I'll keep this open. The `ArgumentNullException` is still a bug; null request bodies should throw something different, or not throw at all.

**Comment by tmenier** [comment]
I decided throw a `FlurlHttpException` with a helpful error message on any attempt to read `HttpCall.RequestBody` when content is not a `CapturedStringContent`. This is now released: https://www.nuget.org/packages/Flurl.Http/1.1.2

**Comment by tmenier** [comment]
Update: Flurl.Http 2.0 will no longer throw on attempted reads of `HttpCall.RequestBody` when content is not a `CapturedStringContent`. It'll just return null. I hope that isn't confusing, but since an `HttpCall` is attached to `FlurlHttpException`, it's very common to inspect it in `catch` blocks, and that's not a good place for something that could unexpectedly throw.
